PYIC-6083: Store inherited identities in evcs

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Store inherited identities in evcs

### Why did it change

[PYIC-6083: Store inherited identities in EVCS](https://github.com/govuk-one-login/ipv-core-back/commit/480522d065f0831ab939755e42b9c6c528eccc68)

This updates the initalise session lambda to store inherited identities
in EVCS, rather than the tactical store.

We're not going to be going live with inherited identities before we're
migrated over to EVCS, so we don't need to worry about migrating
anything.

Adds a new method to the EVCS service to do the updating of an existing
inherited ID to historic, and storing a new one as current.


[PYIC-6083: Do not remove inherited ID when P2 attained](https://github.com/govuk-one-login/ipv-core-back/commit/aa6c604bb4819f527ae4198b9d2cc2805ef8f329)

We no longer want to supersede an inherited identity when a P2
idenentity is achieved - they should be treated separately. This means
not deleting an inherited identity VC if a P2 identity is attained.


[PYIC-6083: Don't update inherited identity when storing P2](https://github.com/govuk-one-login/ipv-core-back/commit/07ae5bd396cc04fdeebcd69019b1f1e141c08583)

We want a user to be able to have an inherited identity along side a P2
identity. This means that if a user with an existing inherited identity
attains a P2, we don't want to lose the inherited identity VC.

This adds metadata to the inherited identity VC in EVCS, which we can
then use when storing completed identities to ensure we don't update the
inherited identity from `CURRENT` to `HISTORIC`.

There are other options for how to do this. We could parse the VCs and
check the issuer, but this is a bit messy and doesn't play nicely with
the streams we're using.

The metadata itself could be in a different shape. This uses the key
"inheritedIdentity" with a value of the Cri ID. This may be more useful
in future that just a boolean flag.

[PYIC-7022: Include current inherited identity VC in bundle](https://github.com/govuk-one-login/ipv-core-back/pull/2223/commits/3cce4ce86f799bb4783499dc211cafc561c42c86)

When we store an inherited identity VC in EVCS it has a status of
`CURRENT`. If a user is returning from a F2F journey we were only using
the VCs with a `PENDING_RETURN` status. We need to ensure we're not
throwing away the current inherited identity VC, so that we can match
the correct VOT in the VTR for inherited identity journeys.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-6083](https://govukverify.atlassian.net/browse/PYIC-6083)


[PYIC-6083]: https://govukverify.atlassian.net/browse/PYIC-6083?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-6083: Store inherited identities in EVCS

This updates the initalise session lambda to store inherited identities
in EVCS, rather than the tactical store.

We're not going to be going live with inherited identities before we're
migrated over to EVCS, so we don't need to worry about migrating
anything.

Adds a new method to the EVCS service to do the updating of an existing
inherited ID to historic, and storing a new one as current.

Wynndow committed:
PYIC-6083: Do not remove inherited ID when P2 attained

We no longer want to supersede an inherited identity when a P2
idenentity is achieved - they should be treated separately. This means
not deleting an inherited identity VC if a P2 identity is attained.

Wynndow committed:
PYIC-6083: Don't update inherited identity when storing P2

We want a user to be able to have an inherited identity along side a P2
identity. This means that if a user with an existing inherited identity
attains a P2, we don't want to lose the inherited identity VC.

This adds metadata to the inherited identity VC in EVCS, which we can
then use when storing completed identities to ensure we don't update the
inherited identity from `CURRENT` to `HISTORIC`.

There are other options for how to do this. We could parse the VCs and
check the issuer, but this is a bit messy and doesn't play nicely with
the streams we're using.

The metadata itself could be in a different shape. This uses the key
"inheritedIdentity" with a value of the Cri ID. This may be more useful
in future that just a boolean flag.

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -321,29 +329,98 @@ private static boolean isListEmpty(List<String> list) {
+
+            evcsService.storeInheritedIdentity(userId, inheritedIdentityVc, null);
+            ipvSessionItem.setInheritedIdentityReceivedThisSession(true);
+            ipvSessionService.updateIpvSession(ipvSessionItem);
```
One concern I've just thought of here... EVCS writes are not synchronous, so we cannot rely on this being immediately readable in `CheckExistingIdentity`.

This might not be easy to solve... We could add it to the IPV Session as well/instead (i.e. `inheritedIdentityReceivedThisSession` goes from a boolean to the VC itself), but then the logic in CheckExistingIdentity becomes fiddly. Perhaps we just park it for now and create a ticket to track later.

Wynndow replied:
Ooo interesting. It would be good to have a proper understanding of what is an acceptable wait between writing and reading. Happy to park it though until we're closer to inherited identities going live

Joe-Edwards-GDS replied:
It's not super precise atm, in 'most' cases it'll be very small, but during their load test it occasionally creeps up to a few seconds.

I think the rule of thumb is that it's probably ok if there's some user interaction between the two, but likely to be a problem if it's all internal processing.

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -321,29 +329,98 @@ private static boolean isListEmpty(List<String> list) {
+        }
+    }
+
+    @Tracing
```
(Do we need tracing on this method? Is it expected to be significant to performance?)

Wynndow replied:
No, I think I was just trying to be consistent - but we don't trace all the private methods in here anyway. Will take it out

Joe-Edwards-GDS replied:
Yeah, I think we want to trace 'operations' that might be useful in diagnosing latency issues, but probably not every private helper method

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -321,29 +329,98 @@ private static boolean isListEmpty(List<String> list) {
+        }
+
+        if (existingInheritedIdentity.isEmpty()) {
+            LOGGER.info("No existing inherited identity found - storing new one");
```
Can we use `LogHelper` when building messages, so we have a consistent JSON structure

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -321,29 +329,98 @@ private static boolean isListEmpty(List<String> list) {
+            throw new InheritedIdentityException(
+                    String.format(
+                            "Too many current inherited identities found - %d",
+                            existingInheritedIdentity.size()));
```
Does this need to be an error?

Wynndow replied:
It's an unexpected situation - and how would you recover from it? Which inherited identity VC from EVCS would you choose to use?

Joe-Edwards-GDS replied:
Yeah, I guess it'd be nice to recover gracefully from it if we can.

In any case we should certainly log a warning, but it feels as though there are two reasonable behaviours afterwards:
- Just handle the first one and ignore any others (if the others are stronger they'll pop up in the matching algorithm if relevant anyway)
- Evaluate whether the new id is weaker than _all_ the existing inherited ids (and mark all of them as historic if they are superseded)

Wynndow replied:
Yeah, that seems reasonable. I think I'll implement option 2 - if the new identity is equal to or stronger than all of them, I'll make them historic

Joe-Edwards-GDS commented on libs/evcs-service/src/main/java/uk/gov/di/ipv/core/library/service/EvcsService.java:
```diff
@@ -220,6 +249,10 @@ private void updateExistingUserVCs(
+                            .filter(
+                                    vc ->
+                                            vc.metadata().get("inheritedIdentity")
+                                                    == null) // Don't update inherited identity VCs
```
Do we need a null-check on `metadata` here?

Wynndow replied:
Probably a good idea

Wynndow committed:
PYIC-7022: Include current inherited identity VC in bundle

When we store an inherited identity VC in EVCS it has a status of
`CURRENT`. If a user is returning from a F2F journey we were only using
the VCs with a `PENDING_RETURN` status. We need to ensure we're not
throwing away the current inherited identity VC, so that we can match
the correct VOT in the VTR for inherited identity journeys.

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -395,6 +395,12 @@ private VerifiableCredentialBundle getEvcsCredentialBundle(
                 evcsIdentityVcs = evcsVcs.get(CURRENT);
                 isPendingEvcs = false;
+            } else {
+                // Ensure we keep any inherited ID VCs in the bundle
```
Good spot! I wonder if, when we revisit inherited ids properly, we should try and store them as a separate status 🤔 - something to worry about later.

Wynndow replied:
This took me a lot longer than I would have liked to find. I find the logic around EVCS and fetching VCs in check-existing-identity hard to grok!

Joe-Edwards-GDS replied:
It's all a little bit fiddly...

Should be simpler once we get rid of the tactical store code, but that's a long way off still.

Joe-Edwards-GDS commented on lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java:
```diff
@@ -321,29 +329,98 @@ private static boolean isListEmpty(List<String> list) {
                     inheritedIdentityVc, auditEventUser, deviceInformation);
-            if (!isHmrcInheritedIdentityWithStrongerVotPresent(inheritedIdentityVc, userId)) {
-                storeHmrcInheritedIdentity(claimsSet, ipvSessionItem, inheritedIdentityVc);
+            if (configService.enabled(EVCS_READ_ENABLED)
```
(Do we need EVCS read enabled here? In practice it won't matter - since we expect inherited ids to come significantly later)

Wynndow replied:
I think I was just being overly cautious. I did contemplate not including any evcs flags here at all and just letting the inherited identity flag do all the work, if we're sure we're not going to use inherited identities before evcs. I'll drop the read check.

Joe-Edwards-GDS reviewed (commented):
Looks like a sensible starting point for inherited identities in EVCS, thanks.

I expect there'll be some rough edges to tidy up when we revisit them 'properly', but that's expected.

Wynndow committed:
PYIC-6083: PR review comments

Wynndow committed:
PYIC-6083: Handle multiple inherited identities

It should never be the case that a user has more than one current
inherited identity in the EVCS. We can handle that scenario gracefully
however if we do encounter it. This will move the existing VCs to
historic if the incoming is stronger or equal to all of them. If it's
weaker we do nothing, and the multiple existing will be used when
matching the requested VOT.

Joe-Edwards-GDS reviewed (dismissed):
Looks good, but you've acquired a conflict with your own PR 😬
